### DISCUSSION TITLE
Montage transparent PNG without losing color information

### DISCUSSION THREAD

**Discussion by bgr** [ImageMagick/ImageMagick#2229] [Help]
Hi,
I need to join 16 RGBA PNGs into a 4x4 grid. The problem I'm having is that _montage_ discards color data where alpha is 0, and I need to retain all four channels as they are in source images.
I've tried numerous combinations of random args I found on the internet (`-channel RGBA`, prepending `PNG32:` to result, some `-alpha` values), nothing helps.

I'm on Windows using latest version of ImageMagick.

**Comment by fmw42** [comment]
You can use +append or +smush to make each row of 4 images. Then use -append or -smush to combine all the rows into the final image

What is your IM version and what was your exact command line?

P.S. I can confirm the issue for IM 6.9.11.22 Q16 Mac OSX with the following commands:

```
convert logo: -transparent white logot.png

montage logot.png logot.png logot.png logot.png \
-channel rgba -alpha on -tile 2x2 -geometry +10+10 \
-define png:color-type=6 PNG32:x.png

```

**Reply by GeeMack** [reply]
Working with fmw42's command above, but using `-background none` to maintain the transparency, you could try something like this...

    convert logo: -transparent white logot.png

    montage logot.png logot.png logot.png logot.png ^
      -background none -tile 2x2 -geometry +10+10 x2.png
 
Can't think of other examples right now, but I come across situations now and then where I think the transparency should stick from an input to an output, but doesn't unless I make sure to add `-background none`.

**Reply by fmw42** [reply]
Nice thought, GeeMack.  It had not occurred to me to try that.

**Comment by snibgo** [comment]
Thanks, GeeMack.

The "-background" setting is used by many operations, such as "-append" when inputs have different widths, "-extent", "-hough-lines", "-layers flatten", "-rotate", "-smush" and "-splice".

I mostly encounter the need for "-background None" in one of the "-layer" operations. I suppose that montage internally does some form of "-layers" operation.

**Comment by bgr** [comment]
Thanks everyone for quick reply, and sorry for not responding sooner. I ended up accomplishing it using 3 commands, but wanted to try out your suggestions before I reply, which I finally got back to.

I'm using Windows build of ImageMagick 7.0.10-22 Q16 x64 2020-06-27.

I have actually been trying using `-background none` before writing my initial message, I should've mentioned it. I couldn't get that to work, RGB data is cleared wherever alpha is 0. This is easy to miss when viewing using image viewer since it's displayed correctly as a transparent image, the problem is evident when channels are inspected individually (e.g. in Photoshop).

I also gave that `-define png:color-type=6 PNG32:x.png` a try now, doesn't seem to make a difference.

The commands I ended up using are:
```
magick montage 0.png 1.png 2.png 3.png -tile 2x2 -mode concatenate -alpha off rgb.png
magick montage 0.png 1.png 2.png 3.png -tile 2x2 -mode concatenate -alpha extract alpha.png
magick rgb.png alpha.png -alpha off -compose CopyOpacity -composite final.png
```
these separately montage RGB and alpha into two separate pngs, then the last command merges them into the final desired image. This solution is sufficient for me, but it'd be good to know it's possible to do it in one line elegantly.

If anyone wants to try to get the one-liner to work, I've uploaded the images I'm using [here](https://www.dropbox.com/s/pc293h2ffcxglt8/imagemagick_montage_alpha.zip?dl=0).
